File: integrations/mcp-use-hol-guard/src/mcp_use_hol_guard/middleware.py
from __future__ import annotations

import asyncio
import inspect
import json
import os
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

from mcp.types import CallToolRequestParams, CallToolResult, TextContent
from mcp_use.middleware import Middleware, MiddlewareContext, NextFunctionT

_MAX_PAYLOAD_BYTES = 24 * 1024
_MAX_OUTPUT_BYTES = 64 * 1024
_DEFAULT_TIMEOUT_SECONDS = 8.0


@dataclass(frozen=True)
class GuardDecision:
    kind: Literal["allow", "deny", "review"]


DecisionProvider = Callable[[dict[str, Any]], GuardDecision | Awaitable[GuardDecision]]


def _object(value: Any) -> dict[str, Any] | None:
    return value if isinstance(value, dict) else None


def _normal(value: Any) -> str | None:
    return value.strip().lower() if isinstance(value, str) and value.strip() else None


def _encode_payload(payload: dict[str, Any]) -> bytes | None:
    try:
        return json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    except (TypeError, ValueError, UnicodeEncodeError):
        return None


def _parse_json_output(raw: bytes) -> dict[str, Any] | None:
    if len(raw) > _MAX_OUTPUT_BYTES:
        return None
    try:
        text = raw.decode("utf-8", errors="strict").strip()
    except UnicodeDecodeError:
        return None
    if not text:
        return None
    for candidate in [text, *reversed(text.splitlines())]:
        try:
            parsed = json.loads(candidate)
        except json.JSONDecodeError:
            continue
        if isinstance(parsed, dict):
            return parsed
    return None


def _classify(payload: dict[str, Any] | None) -> GuardDecision | None:
    if payload is None:
        return None
    queue = [payload]
    seen: set[int] = set()
    allow = False
    review = False
    deny = False
    while queue:
        if len(seen) >= 32:
            return None
        layer = queue.pop(0)
        marker = id(layer)
        if marker in seen:
            return None
        seen.add(marker)
        hook = _object(layer.get("hookSpecificOutput")) or {}
        decisions = {
            value
            for value in (
                _normal(hook.get("permissionDecision")),
                _normal(layer.get("permissionDecision")),
                _normal(layer.get("decision")),
            )
            if value
        }
        actions = {
            value
            for value in (_normal(layer.get("policy_action")), _normal(layer.get("policyAction")))
            if value
        }
        deny = deny or layer.get("blocked") is True or layer.get("continue") is False
        deny = deny or bool(decisions.intersection({"deny", "block"}))
        deny = deny or bool(actions.intersection({"block", "sandbox-required"}))
        review = review or bool(decisions.intersection({"review", "ask"}))
        review = review or bool(actions.intersection({"review", "require-reapproval"}))
        allow = allow or "allow" in decisions or "allow" in actions
        for key in ("data", "payload", "result"):
            nested = _object(layer.get(key))
            if nested is not None:
                queue.append(nested)
    if deny:
        return GuardDecision("deny")
    if review:
        return GuardDecision("review")
    if allow:
        return GuardDecision("allow")
    return None


async def _terminate_process(process: asyncio.subprocess.Process) -> None:
    if process.returncode is None:
        try:
            process.kill()
        except ProcessLookupError:
            pass
    try:
        await process.wait()
    except ProcessLookupError:
        pass


async def _communicate_bounded(process: asyncio.subprocess.Process, encoded: bytes) -> bytes | None:
    if process.stdin is None or process.stdout is None:
        await _terminate_process(process)
        return None
    try:
        process.stdin.write(encoded)
        await process.stdin.drain()
    except (BrokenPipeError, ConnectionResetError):
        await _terminate_process(process)
        return None
    finally:
        process.stdin.close()

    chunks: list[bytes] = []
    total = 0
    while True:
        read_limit = min(8192, _MAX_OUTPUT_BYTES + 1 - total)
        chunk = await process.stdout.read(read_limit)
        if not chunk:
            break
        total += len(chunk)
        if total > _MAX_OUTPUT_BYTES:
            await _terminate_process(process)
            return None
        chunks.append(chunk)
    await process.wait()
    return b"".join(chunks)


class LocalHOLGuardProvider:
    def __init__(
        self,
        *,
        command: str = "hol-guard",
        workspace: str | os.PathLike[str] | None = None,
        timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
    ) -> None:
        self.command = command
        self.workspace = Path(workspace or os.environ.get("HOL_GUARD_WORKSPACE") or os.getcwd()).resolve()
        self.timeout_seconds = timeout_seconds

    async def __call__(self, payload: dict[str, Any]) -> GuardDecision:
        encoded = _encode_payload(payload)
        if encoded is None or len(encoded) > _MAX_PAYLOAD_BYTES:
            return GuardDecision("deny")
        process: asyncio.subprocess.Process | None = None
        try:
            process = await asyncio.create_subprocess_exec(
                self.command,
                "guard",
                "hook",
                "--harness",
                "generic",
                "--workspace",
                str(self.workspace),
                cwd=self.workspace,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.DEVNULL,
            )
            stdout = await asyncio.wait_for(_communicate_bounded(process, encoded), timeout=self.timeout_seconds)
        except (OSError, asyncio.TimeoutError):
            if process is not None:
                await _terminate_process(process)
            return GuardDecision("deny")
        decision = _classify(_parse_json_output(stdout)) if stdout is not None else None
        if process.returncode != 0 or decision is None:
            return GuardDecision("deny")
        return decision


def _blocked_result(message: str) -> CallToolResult:
    return CallToolResult(isError=True, content=[TextContent(type="text", text=message)])


class MCPUseHOLGuardMiddleware(Middleware):
    """Gate mcp-use tools/call requests before the downstream MCP server is invoked."""

    def __init__(self, provider: DecisionProvider | None = None, *, workspace: str | os.PathLike[str] | None = None) -> None:
        self._provider = provider or LocalHOLGuardProvider(workspace=workspace)

    async def on_call_tool(
        self,
        context: MiddlewareContext[CallToolRequestParams],
        call_next: NextFunctionT,
    ) -> CallToolResult:
        arguments = context.params.arguments or {}
        payload = {
            "hook_event_name": "PreToolUse",
            "hookEventName": "PreToolUse",
            "tool_name": context.params.name,
            "tool_input": arguments,
            "cwd": str(getattr(self._provider, "workspace", Path.cwd())),
            "runtime_context": {
                "framework": "mcp-use",
                "request_id": context.id,
                "connection_id": context.connection_id,
            },
        }
        encoded = _encode_payload(payload)
        if encoded is None:
            return _blocked_result("HOL Guard could not validate this MCP tool request.")
        if len(encoded) > _MAX_PAYLOAD_BYTES:
            return _blocked_result("HOL Guard rejected an oversized MCP tool request.")
        try:
            value = self._provider(payload)
            decision = await value if inspect.isawaitable(value) else value
        except Exception:
            decision = GuardDecision("deny")
        if not isinstance(decision, GuardDecision) or decision.kind not in {"allow", "deny", "review"}:
            decision = GuardDecision("deny")
        if decision.kind == "allow":
            return await call_next(context)
        if decision.kind == "review":
            return _blocked_result("HOL Guard requires approval before this MCP tool can run.")
        return _blocked_result("HOL Guard blocked this MCP tool call.")

File: integrations/mcp-sdk-go-hol-guard/middleware.go
package holguardmcpsdk

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

const (
	maxPayloadBytes        = 24 * 1024
	maxDecisionOutputBytes = 64 * 1024
)

type Action string

const (
	ActionAllow  Action = "allow"
	ActionDeny   Action = "deny"
	ActionReview Action = "review"
)

type Decision struct {
	Action Action
}

type ToolCall struct {
	Name      string
	Arguments map[string]any
}

type DecisionProvider interface {
	Evaluate(context.Context, ToolCall) (Decision, error)
}

type DecisionProviderFunc func(context.Context, ToolCall) (Decision, error)

func (f DecisionProviderFunc) Evaluate(ctx context.Context, call ToolCall) (Decision, error) {
	return f(ctx, call)
}

// ReceivingMiddleware gates inbound tools/call requests before the registered
// MCP tool handler executes. All non-tool methods pass through unchanged.
func ReceivingMiddleware(provider DecisionProvider) mcp.Middleware {
	if provider == nil {
		panic("holguardmcpsdk: nil decision provider")
	}
	return func(next mcp.MethodHandler) mcp.MethodHandler {
		return func(ctx context.Context, method string, req mcp.Request) (mcp.Result, error) {
			toolReq, ok := req.(*mcp.CallToolRequest)
			if !ok {
				return next(ctx, method, req)
			}

			name := strings.TrimSpace(toolReq.Params.Name)
			if name == "" {
				return denyResult("HOL Guard could not identify the MCP tool call"), nil
			}

			if len(toolReq.Params.Arguments) > maxPayloadBytes {
				return denyResult("HOL Guard could not safely evaluate MCP tool arguments"), nil
			}
			arguments := map[string]any{}
			if len(toolReq.Params.Arguments) > 0 {
				if err := json.Unmarshal(toolReq.Params.Arguments, &arguments); err != nil {
					return denyResult("HOL Guard could not safely evaluate MCP tool arguments"), nil
				}
			}

			decision, err := evaluateSafely(provider, ctx, ToolCall{Name: name, Arguments: arguments})
			if err != nil {
				return denyResult("HOL Guard was unavailable; MCP tool execution failed closed"), nil
			}
			switch decision.Action {
			case ActionAllow:
				return next(ctx, method, req)
			case ActionDeny:
				return denyResult("HOL Guard denied MCP tool execution"), nil
			case ActionReview:
				return denyResult("HOL Guard requires approval before MCP tool execution"), nil
			default:
				return denyResult("HOL Guard returned no authoritative decision; MCP tool execution failed closed"), nil
			}
		}
	}
}

func denyResult(message string) *mcp.CallToolResult {
	return &mcp.CallToolResult{
		IsError: true,
		Content: []mcp.Content{&mcp.TextContent{Text: message}},
	}
}

func evaluateSafely(provider DecisionProvider, ctx context.Context, call ToolCall) (decision Decision, err error) {
	defer func() {
		if recover() != nil {
			decision = Decision{}
			err = errors.New("HOL Guard decision provider panicked")
		}
	}()
	return provider.Evaluate(ctx, call)
}

type LocalProvider struct {
	Executable string
	Workspace  string
	Timeout    time.Duration
}

func (p LocalProvider) Evaluate(ctx context.Context, call ToolCall) (Decision, error) {
	executable := strings.TrimSpace(p.Executable)
	if executable == "" {
		executable = "hol-guard"
	}
	timeout := p.Timeout
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	if timeout < 250*time.Millisecond || timeout > 10*time.Second {
		return Decision{}, errors.New("invalid HOL Guard timeout")
	}
	workspace := strings.TrimSpace(p.Workspace)
	if workspace != "" {
		var err error
		workspace, err = filepath.Abs(workspace)
		if err != nil {
			return Decision{}, errors.New("invalid workspace")
		}
	}

	payload := map[string]any{
		"hook_event_name": "PreToolUse",
		"tool_name":       call.Name,
		"tool_input":      call.Arguments,
		"framework":       "mcp-go-sdk",
		"source_scope":    map[bool]string{true: "project", false: "global"}[workspace != ""],
	}
	encoded, err := json.Marshal(payload)
	if err != nil || len(encoded) > maxPayloadBytes {
		return Decision{}, errors.New("HOL Guard request exceeds adapter limit")
	}

	decisionCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	argv := []string{"guard", "hook", "--harness", "generic"}
	if workspace != "" {
		argv = append(argv, "--workspace", workspace)
	}
	argv = append(argv, "--json")
	cmd := exec.CommandContext(decisionCtx, executable, argv...)
	cmd.Stdin = bytes.NewReader(encoded)
	cmd.Stderr = io.Discard
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return Decision{}, errors.New("HOL Guard decision process could not start")
	}
	if err := cmd.Start(); err != nil {
		return Decision{}, errors.New("HOL Guard decision process could not start")
	}
	output, readErr := io.ReadAll(io.LimitReader(stdout, maxDecisionOutputBytes+1))
	if readErr != nil {
		cancel()
		_ = cmd.Wait()
		return Decision{}, errors.New("HOL Guard decision output could not be read")
	}
	if len(output) > maxDecisionOutputBytes {
		cancel()
		_ = cmd.Wait()
		return Decision{}, errors.New("HOL Guard decision output exceeded adapter limit")
	}
	if err := cmd.Wait(); err != nil {
		return Decision{}, errors.New("HOL Guard decision process failed")
	}
	return parseDecision(output)
}

func parseDecision(output []byte) (Decision, error) {
	lines := bytes.Split(bytes.TrimSpace(output), []byte("\n"))
	for i := len(lines) - 1; i >= 0; i-- {
		line := bytes.TrimSpace(lines[i])
		if len(line) == 0 || line[0] != '{' {
			continue
		}
		var payload map[string]any
		if json.Unmarshal(line, &payload) != nil {
			continue
		}
		if blocked, _ := payload["blocked"].(bool); blocked {
			return Decision{Action: ActionDeny}, nil
		}
		if cont, ok := payload["continue"].(bool); ok && !cont {
			return Decision{Action: ActionDeny}, nil
		}
		for _, key := range []string{"policy_action", "policyAction", "decision", "permissionDecision"} {
			if value, ok := payload[key].(string); ok {
				if d, ok := normalizeDecision(value); ok {
					return d, nil
				}
			}
		}
		if hook, ok := payload["hookSpecificOutput"].(map[string]any); ok {
			if value, ok := hook["permissionDecision"].(string); ok {
				if d, ok := normalizeDecision(value); ok {
					return d, nil
				}
			}
	}
	return Decision{}, errors.New("HOL Guard returned no authoritative decision")
}

func normalizeDecision(value string) (Decision, bool) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "allow", "warn":
		return Decision{Action: ActionAllow}, true
	case "ask", "review", "require-reapproval":
		return Decision{Action: ActionReview}, true
	case "deny", "block", "sandbox-required":
		return Decision{Action: ActionDeny}, true
	default:
		return Decision{}, false
	}
}
